Readme.md
NBT.js [![Build Status](https://travis-ci.org/sjmulder/nbt-js.png)](https://travis-ci.org/sjmulder/nbt-js)
======

By Sijmen Mulder and a host of wonderful contributors.

NBT.js is a JavaScript parser and serializer for [NBT](http://wiki.vg/NBT)
archives, for use with [Node.js](http://nodejs.org/) or the browser.


Usage
-----

After `var nbt = require('nbt')` or `<script src="nbt.js"></script>`, you can
use `nbt.parse(data, callback)` to convert NBT data into a regular JavaScript
object.

```js
var fs = require('fs'),
    nbt = require('nbt');

var data = fs.readFileSync('fixtures/bigtest.nbt.gz');
nbt.parse(data, function(error, data) {
    if (error) { throw error; }

    console.log(data.value.stringTest.value);
    console.log(data.value['nested compound test'].value);
});
```

If the data is gzipped, it is automatically decompressed first. When running
in the browser, `window.zlib` is required for this to work.

Tag names are copied verbatim, and as some names are not valid JavaScript
names, use of the indexer may be required – such as with the nested
compound test in the example above.


API documentation
-----------------

The full documentation generated with JSDoc is available in the docs/
directory and online:

http://sjmulder.github.io/nbt-js/


Development and testing
-----------------------

```bash
npm install  # Install development dependencies
make check   # Check code quality with jshint and run tests
make watch   # Automatically runs 'make check' every few seconds
make doc     # Regenerate the documentation in docs/
```

Copyright
---------

I, the copyright holder of this work, hereby release it into the public
domain. This applies worldwide.

In case this is not legally possible: I grant anyone the right to use this
work for any purpose, without any conditions, unless such conditions are
required by law.


Changelog.md
Changelog
=========

0.8.1 (2018-12-16)
-----------------

 * Fixed version in package.json

0.8.0 (2018-12-16)
------------------

 * New: longArray data type (thanks sourcehunter).
 * Change: dropped linting tools and PhantomJS tests.

0.7.0
-----

 * Feature: browser support, simply include nbt.js as-is.
 * Feature: new Reader/Writer.ubyte() for unsigned bytes.
 * Improvement: JSDoc documentation added.
 * Improvement: updated dependencies and upgraded to Node 6.
 * Improvement: simplified build by using make instead of grunt.
 * Improvement: added .editorconfig.
 * Improvement: automated code style checking with eclint.
 * Improvement: automatic NPM deployments with Travis CI.

0.6.0
-----

The big change here is that `parse()` now returns a consistent format again:

    parse(foo); // -> { name: 'Level', value: { ... } }

Similarly, that's what `writeUncompressed()` as well now:

    writeUncompressed({ name: 'Level', value: { ... } });

 * Improvement: dropped unused jasmine package dependency.
 * Improvement: [#13] read() and write() cycles are now guaranteed to be
   idempotent.
 * Improvement: [#14] back to consistent structure for parse() result.
 * Improvement: [#15] `writer.byteArray()` now accepts both arrays and
   `Buffer`s.
 * Fix: parse() would always throw when used on non-gzipped data.

0.5.0
-----

A merge and slight rework of the terrific improvements by the people at
[ProsmarineJS](https://github.com/PrismarineJS/prismarine-nbt).

 * Feature: support for writing archives (to a `Buffer`).
 * Improvement: switch to Mocha testing framwork.
 * Improvement: `Reader.offset` is now exposed.
 * Improvement: 64 bit values now represented as [upper, lower] pairs for
   better portability.
 * Improvement: exposed `nbt.writeUncompressed`, which is sync as opposed
   to its unzipping counterpart.
 * Improvement: lists and compounds now return type information. This creates
   a symetry because this output can be fed into the writer as-is.

0.4.0
-----

 * Feature: the `Reader` constructor is now public and can be used eo read
   individual NBT elements from a buffer.
 * Feature: support for `npm test` using the previously built testing
   infrastructure.
 * Improvement: greatly expanded test coverage.
 * Improvement: byte array values are now returned as arrays rather than
   Buffer objects.
 * Improvement: continuous integration with Travis now runs test against the
   GitHub repository.
 * Improvement: added a changelog.
 * Fix: the long reader returned incorrect values due to an operator
   precedence issue.

0.3.0
-----

 * Improvement: jshint Grunt task for checking code quality.
 * Improvement: small fixes throughout based on jshint tips.
 * Improvement: added various unit tests with Jasmine.
 * Improvement: added links in documentation for easy lookup.
 * Improvement: dropped bundled binary library for built-in `Buffer`.

0.2.0
-----

 * Feature: support for gzipped archives.
 * Feature: support for int arrays (type 11).
 * Improvement: when the root node is a compound with no name, only the value
   is returned.
 * Improvement: better documentation.
 * Improvement: when an error occurs, an Error is thrown rather than as
   string.

0.1.0
-----

Initial release.


sample/bigtest.js
module.exports = {
    "name": "Level",
    "value": {
        "longTest": {
            "type": "long",
            "value": [
                2147483647,
                -1
            ]
        },
        "shortTest": {
            "type": "short",
            "value": 32767
        },
        "stringTest": {
            "type": "string",
            "value": "HELLO WORLD THIS IS A TEST STRING ÅÄÖ!"
        },
        "floatTest": {
            "type": "float",
            "value": 0.4982314705848694
        },
        "intTest": {
            "type": "int",
            "value": 2147483647
        },
        "nested compound test": {
            "type": "compound",
            "value": {
                "ham": {
                    "type": "compound",
                    "value": {
                        "name": {
                            "type": "string",
                            "value": "Hampus"
                        },
                        "value": {
                            "type": "float",
                            "value": 0.75
                        }
                    }
                },
                "egg": {
                    "type": "compound",
                    "value": {
                        "name": {
                            "type": "string",
                            "value": "Eggbert"
                        },
                        "value": {
                            "type": "float",
                            "value": 0.5
                        }
                    }
                }
            }
        },
        "listTest (long)": {
            "type": "list",
            "value": {
                "type": "long",
                "value": [
                    [
                        0,
                        11
                    ],
                    [
                        0,
                        12
                    ],
                    [
                        0,
                        13
                    ],
                    [
                        0,
                        14
                    ],
                    [
                        0,
                        15
                    ]
                ]
            }
        },
        "listTest (compound)": {
            "type": "list",
            "value": {
                "type": "compound",
                "value": [
                    {
                        "name": {
                            "type": "string",
                            "value": "Compound tag #0"
                        },
                        "created-on": {
                            "type": "long",
                            "value": [
                                294,
                                1379390861
                            ]
                        }
                    },
                    {
                        "name": {
                            "type": "string",
                            "value": "Compound tag #1"
                        },
                        "created-on": {
                            "type": "long",
                            "value": [
                                294,
                                1379390861
                            ]
                        }
                    }
                ]
            }
        },
        "byteTest": {
            "type": "byte",
            "value": 127
        },
        "byteArrayTest (the first 1000 values of (n*n*255+n*7)%100, starting with n=0 (0, 62, 34, 16, 8, ...))": {
            "type": "byteArray",
            "value": [
                0,
                62,
                34,
                16,
                8,
                10,
                22,
                44,
                76,
                18,
                70,
                32,
                4,
                86,
                78,
                80,
                92,
                14,
                46,
                88,
                40,
                2,
                74,
                56,
                48,
                50,
                62,
                84,
                16,
                58,
                10,
                72,
                44,
                26,
                18,
                20,
                32,
                54,
                86,
                28,
                80,
                42,
                14,
                96,
                88,
                90,
                2,
                24,
                56,
                98,
                50,
                12,
                84,
                66,
                58,
                60,
                72,
                94,
                26,
                68,
                20,
                82,
                54,
                36,
                28,
                30,
                42,
                64,
                96,
                38,
                90,
                52,
                24,
                6,
                98,
                0,
                12,
                34,
                66,
                8,
                60,
                22,
                94,
                76,
                68,
                70,
                82,
                4,
                36,
                78,
                30,
                92,
                64,
                46,
                38,
                40,
                52,
                74,
                6,
                48,
                0,
                62,
                34,
                16,
                8,
                10,
                22,
                44,
                76,
                18,
                70,
                32,
                4,
                86,
                78,
                80,
                92,
                14,
                46,
                88,
                40,
                2,
                74,
                56,
                48,
                50,
                62,
                84,
                16,
                58,
                10,
                72,
                44,
                26,
                18,
                20,
                32,
                54,
                86,
                28,
                80,
                42,
                14,
                96,
                88,
                90,
                2,
                24,
                56,
                98,
                50,
                12,
                84,
                66,
                58,
                60,
                72,
                94,
                26,
                68,
                20,
                82,
                54,
                36,
                28,
                30,
                42,
                64,
                96,
                38,
                90,
                52,
                24,
                6,
                98,
                0,
                12,
                34,
                66,
                8,
                60,
                22,
                94,
                76,
                68,
                70,
                82,
                4,
                36,
                78,
                30,
                92,
                64,
                46,
                38,
                40,
                52,
                74,
                6,
                48,
                0,
                62,
                34,
                16,
                8,
                10,
                22,
                44,
                76,
                18,
                70,
                32,
                4,
                86,
                78,
                80,
                92,
                14,
                46,
                88,
                40,
                2,
                74,
                56,
                48,
                50,
                62,
                84,
                16,
                58,
                10,
                72,
                44,
                26,
                18,
                20,
                32,
                54,
                86,
                28,
                80,
                42,
                14,
                96,
                88,
                90,
                2,
                24,
                56,
                98,
                50,
                12,
                84,
                66,
                58,
                60,
                72,
                94,
                26,
                68,
                20,
                82,
                54,
                36,
                28,
                30,
                42,
                64,
                96,
                38,
                90,
                52,
                24,
                6,
                98,
                0,
                12,
                34,
                66,
                8,
                60,
                22,
                94,
                76,
                68,
                70,
                82,
                4,
                36,
                78,
                30,
                92,
                64,
                46,
                38,
                40,
                52,
                74,
                6,
                48,
                0,
                62,
                34,
                16,
                8,
                10,
                22,
                44,
                76,
                18,
                70,
                32,
                4,
                86,
                78,
                80,
                92,
                14,
                46,
                88,
                40,
                2,
                74,
                56,
                48,
                50,
                62,
                84,
                16,
                58,
                10,
                72,
                44,
                26,
                18,
                20,
                32,
                54,
                86,
                28,
                80,
                42,
                14,
                96,
                88,
                90,
                2,
                24,
                56,
                98,
                50,
                12,
                84,
                66,
                58,
                60,
                72,
                94,
                26,
                68,
                20,
                82,
                54,
                36,
                28,
                30,
                42,
                64,
                96,
                38,
                90,
                52,
                24,
                6,
                98,
                0,
                12,
                34,
                66,
                8,
                60,
                22,
                94,
                76,
                68,
                70,
                82,
                4,
                36,
                78,
                30,
                92,
                64,
                46,
                38,
                40,
                52,
                74,
                6,
                48,
                0,
                62,
                34,
                16,
                8,
                10,
                22,
                44,
                76,
                18,
                70,
                32,
                4,
                86,
                78,
                80,
                92,
                14,
                46,
                88,
                40,
                2,
                74,
                56,
                48,
                50,
                62,
                84,
                16,
                58,
                10,
                72,
                44,
                26,
                18,
                20,
                32,
                54,
                86,
                28,
                80,
                42,
                14,
                96,
                88,
                90,
                2,
                24,
                56,
                98,
                50,
                12,
                84,
                66,
                58,
                60,
                72,
                94,
                26,
                68,
                20,
                82,
                54,
                36,
                28,
                30,
                42,
                64,
                96,
                38,
                90,
                52,
                24,
                6,
                98,
                0,
                12,
                34,
                66,
                8,
                60,
                22,
                94,
                76,
                68,
                70,
                82,
                4,
                36,
                78,
                30,
                92,
                64,
                46,
                38,
                40,
                52,
                74,
                6,
                48,
                0,
                62,
                34,
                16,
                8,
                10,
                22,
                44,
                76,
                18,
                70,
                32,
                4,
                86,
                78,
                80,
                92,
                14,
                46,
                88,
                40,
                2,
                74,
                56,
                48,
                50,
                62,
                84,
                16,
                58,
                10,
                72,
                44,
                26,
                18,
                20,
                32,
                54,
                86,
                28,
                80,
                42,
                14,
                96,
                88,
                90,
                2,
                24,
                56,
                98,
                50,
                12,
                84,
                66,
                58,
                60,
                72,
                94,
                26,
                68,
                20,
                82,
                54,
                36,
                28,
                30,
                42,
                64,
                96,
                38,
                90,
                52,
                24,
                6,
                98,
                0,
                12,
                34,
                66,
                8,
                60,
                22,
                94,
                76,
                68,
                70,
                82,
                4,
                36,
                78,
                30,
                92,
                64,
                46,
                38,
                40,
                52,
                74,
                6,
                48,
                0,
                62,
                34,
                16,
                8,
                10,
                22,
                44,
                76,
                18,
                70,
                32,
                4,
                86,
                78,
                80,
                92,
                14,
                46,
                88,
                40,
                2,
                74,
                56,
                48,
                50,
                62,
                84,
                16,
                58,
                10,
                72,
                44,
                26,
                18,
                20,
                32,
                54,
                86,
                28,
                80,
                42,
                14,
                96,
                88,
                90,
                2,
                24,
                56,
                98,
                50,
                12,
                84,
                66,
                58,
                60,
                72,
                94,
                26,
                68,
                20,
                82,
                54,
                36,
                28,
                30,
                42,
                64,
                96,
                38,
                90,
                52,
                24,
                6,
                98,
                0,
                12,
                34,
                66,
                8,
                60,
                22,
                94,
                76,
                68,
                70,
                82,
                4,
                36,
                78,
                30,
                92,
                64,
                46,
                38,
                40,
                52,
                74,
                6,
                48,
                0,
                62,
                34,
                16,
                8,
                10,
                22,
                44,
                76,
                18,
                70,
                32,
                4,
                86,
                78,
                80,
                92,
                14,
                46,
                88,
                40,
                2,
                74,
                56,
                48,
                50,
                62,
                84,
                16,
                58,
                10,
                72,
                44,
                26,
                18,
                20,
                32,
                54,
                86,
                28,
                80,
                42,
                14,
                96,
                88,
                90,
                2,
                24,
                56,
                98,
                50,
                12,
                84,
                66,
                58,
                60,
                72,
                94,
                26,
                68,
                20,
                82,
                54,
                36,
                28,
                30,
                42,
                64,
                96,
                38,
                90,
                52,
                24,
                6,
                98,
                0,
                12,
                34,
                66,
                8,
                60,
                22,
                94,
                76,
                68,
                70,
                82,
                4,
                36,
                78,
                30,
                92,
                64,
                46,
                38,
                40,
                52,
                74,
                6,
                48,
                0,
                62,
                34,
                16,
                8,
                10,
                22,
                44,
                76,
                18,
                70,
                32,
                4,
                86,
                78,
                80,
                92,
                14,
                46,
                88,
                40,
                2,
                74,
                56,
                48,
                50,
                62,
                84,
                16,
                58,
                10,
                72,
                44,
                26,
                18,
                20,
                32,
                54,
                86,
                28,
                80,
                42,
                14,
                96,
                88,
                90,
                2,
                24,
                56,
                98,
                50,
                12,
                84,
                66,
                58,
                60,
                72,
                94,
                26,
                68,
                20,
                82,
                54,
                36,
                28,
                30,
                42,
                64,
                96,
                38,
                90,
                52,
                24,
                6,
                98,
                0,
                12,
                34,
                66,
                8,
                60,
                22,
                94,
                76,
                68,
                70,
                82,
                4,
                36,
                78,
                30,
                92,
                64,
                46,
                38,
                40,
                52,
                74,
                6,
                48,
                0,
                62,
                34,
                16,
                8,
                10,
                22,
                44,
                76,
                18,
                70,
                32,
                4,
                86,
                78,
                80,
                92,
                14,
                46,
                88,
                40,
                2,
                74,
                56,
                48,
                50,
                62,
                84,
                16,
                58,
                10,
                72,
                44,
                26,
                18,
                20,
                32,
                54,
                86,
                28,
                80,
                42,
                14,
                96,
                88,
                90,
                2,
                24,
                56,
                98,
                50,
                12,
                84,
                66,
                58,
                60,
                72,
                94,
                26,
                68,
                20,
                82,
                54,
                36,
                28,
                30,
                42,
                64,
                96,
                38,
                90,
                52,
                24,
                6,
                98,
                0,
                12,
                34,
                66,
                8,
                60,
                22,
                94,
                76,
                68,
                70,
                82,
                4,
                36,
                78,
                30,
                92,
                64,
                46,
                38,
                40,
                52,
                74,
                6,
                48
            ]
        },
        "doubleTest": {
            "type": "double",
            "value": 0.4931287132182315
        }
    }
};
